pkg/auth/metadata.go

```diff
@@ -16,6 +16,12 @@ const (
 	SetupTimestampKeyName = "setup_timestamp"
 )
 
+//nolint:gochecknoinits
+func init() {
+	kv.RegisterType(InstallationIDKeyName, nil)
+	kv.RegisterType(SetupTimestampKeyName, nil)
```

Comment:
I hope you meant lakefs,right?
Eventually, if I want the `kv` tool to be able to parse the value from KV correctly, it has to be familiar with the type - am I wrong?

Reply:
As I see it - we should be able to describe all message types using the right partition/key patterns. I understand that the current state prevent this from happening, and I think it can cause conflcits in case we render, for example, staging token and branch information which both use custom paritition keys with path that can be any object that cloud be match a branch name.

Without adding any optimization I've used the following to register all the types:

```go
var recordsMatches []MatchRecord

// RegisterType - Register a pb message type to parse the data, according to a path regex
// All objects which match the path regex will be parsed as that type
// A nil type parses the value as a plain string
func RegisterType(partitionPattern, pathPattern string, mt protoreflect.MessageType) {
	recordsMatches = append(recordsMatches, MatchRecord{
		PartitionPattern: partitionPattern,
		PathPattern:      pathPattern,
		MessageType:      mt,
	})
	sort.Slice(recordsMatches, func(i, j int) bool {
		p1 := strings.TrimPrefix(recordsMatches[i].PartitionPattern, "*")
		p2 := strings.TrimPrefix(recordsMatches[j].PartitionPattern, "*")
		if p1 == p2 {
			pp1 := strings.TrimPrefix(recordsMatches[i].PathPattern, "*")
			pp2 := strings.TrimPrefix(recordsMatches[j].PathPattern, "*")
			return pp1 > pp2
		}
		return p1 > p2
	})
}
```

Used the following to create a record (previously kv object):

```go
func NewRecord(partition, path string, rawValue []byte) (*Record, error) {
	var mt protoreflect.MessageType
	for _, r := range recordsMatches {
		partitionMatch, err := filepath.Match(r.PartitionPattern, partition)
		if err != nil {
			return nil, err
		}
		if !partitionMatch {
			continue
		}

		pathMatch, err := filepath.Match(r.PathPattern, path)
		if err != nil {
			return nil, err
		}
		if pathMatch {
			mt = r.MessageType
			break
		}
	}

	var value interface{}
	if mt == nil {
		value = rawValue
	} else {
		msg := mt.New().Interface()
		err := proto.Unmarshal(rawValue, msg)
		if err != nil {
			return nil, err
		}
		value = msg
	}

	return &Record{
		Partition: partition,
		Key:       path,
		Value:     value,
	}, nil
}
```

Having each message type register using pattern matching:

```go
	kv.RegisterType(PartitionKey, "repos/*/tasks/*", (&TaskResultData{}).ProtoReflect().Type())
	kv.RegisterType(PartitionKey, "repos/*/runs/*", (&RunResultData{}).ProtoReflect().Type())
	kv.RegisterType(PartitionKey, "repos/*/branches/*/*", (&kv.SecondaryIndex{}).ProtoReflect().Type())
	kv.RegisterType(PartitionKey, "repos/*/commits/*/*", (&kv.SecondaryIndex{}).ProtoReflect().Type())
	//...
	kv.RegisterType(PartitionKey, "users/*", (&UserData{}).ProtoReflect().Type())
	kv.RegisterType(PartitionKey, "policies/*", (&PolicyData{}).ProtoReflect().Type())
	kv.RegisterType(PartitionKey, "groups/*", (&GroupData{}).ProtoReflect().Type())
	kv.RegisterType(PartitionKey, "uCredentials/*/credentials/*", (&CredentialData{}).ProtoReflect().Type())
	kv.RegisterType(PartitionKey, "gUsers/*/users/*", (&kv.SecondaryIndex{}).ProtoReflect().Type())
	kv.RegisterType(PartitionKey, "gPolicies/*/policies/*", (&kv.SecondaryIndex{}).ProtoReflect().Type())
	kv.RegisterType(PartitionKey, "uPolicies/*/policies/*", (&kv.SecondaryIndex{}).ProtoReflect().Type())
	kv.RegisterType(PartitionKey, "expiredTokens/*", (&TokenData{}).ProtoReflect().Type())
	kv.RegisterType(PartitionKey, "installation_metadata/*", nil)
	//...
	kv.RegisterType(gravelerPartition, "repos/*", (&RepositoryData{}).ProtoReflect().Type())
	kv.RegisterType("*", "branches/*", (&BranchData{}).ProtoReflect().Type())
	kv.RegisterType("*", "commits/*", (&CommitData{}).ProtoReflect().Type())
	kv.RegisterType("*", "tags/*", (&TagData{}).ProtoReflect().Type())
	kv.RegisterType("*", "*", (&StagedEntryData{}).ProtoReflect().Type())
```

The above support lookup based on partition key which make the result expected - not based on the un-ordered map.
Used filepath match to enabled fast matching - not match regex pattern at runtime which is very slow.
Used strings - but we can format the path using `kv.FormatPath`.
Don't use 'default message' as we have fallback by using '*' '*.

Reply:
Thanks for this suggestion. Decided to take it as is. Regex usage is removed all together